int32 p_x
int32 p_y
int32 p_z